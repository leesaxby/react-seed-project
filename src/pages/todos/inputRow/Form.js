import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup, FormControl } from 'react-bootstrap';
import { injectIntl, intlShape, defineMessages } from 'react-intl';

const messages = defineMessages({
    'todo.form.addItem': {
        id: 'todo.form.addItem',
        description: 'Placeholder for the todo input',
        defaultMessage: 'Add Item'
    }
});

class Form extends React.Component {

    static propTypes = {
        onAddTodoItem: PropTypes.func.isRequired,
        intl: intlShape.isRequired,
    };

    state = {
        newItem: ''
    };

    render() {
        return (
            <form onSubmit={this.submitTodo}>
                <FormGroup bsSize="large">
                    <FormControl type="text"
                                 data-test-id="todo-add-item"
                                 placeholder={this.props.intl.formatMessage(messages['todo.form.addItem'])}
                                 aria-label={this.props.intl.formatMessage(messages['todo.form.addItem'])}
                                 value={this.state.newItem}
                                 onChange={this.updateNewItem}
                                 onKeyPress={this.handleKeyPress}/>
                </FormGroup>
            </form>
        );
    }

    addTodoItem = () => {
        this.props.onAddTodoItem(this.state.newItem);
        this.setState({ newItem: '' });
    }

    updateNewItem = (e) => {
        this.setState({ newItem: e.target.value });
    }

    submitTodo = (e) => {
        this.addTodoItem();
        e.preventDefault();
    }
    
}

export default injectIntl(Form);