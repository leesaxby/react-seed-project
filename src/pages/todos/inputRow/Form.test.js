import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { FormControl } from 'react-bootstrap';
import { Form } from './Form';

describe('A Form', () => {

	function setup(mountFunc = shallow) {
		const props = {
		    onAddItem: jest.fn(),
		    intl: {
		      formatMessage: jest.fn(),
		    },
		};
		const wrapper = mountFunc(<Form { ...props } />);

		return {
			props,
			wrapper,
		};
	}

	it('renders component', () => {
		const { props } = setup();
		const tree = renderer.create(
			<Form { ...props } />
		).toJSON();

		expect(tree).toMatchSnapshot();
	});

   it('keeps track of what is being entered in the input box', () => {
    	const { wrapper } = setup();

    	expect(wrapper.state('newItem')).toBe('');

    	const formControl = wrapper.find(FormControl).first();
    	formControl.simulate('change', { target: { value: 'changed' } } );
    	expect(wrapper.state('newItem')).toBe('changed');
    });

    it('triggers a callback when the input box is submitted', () => {
    	const { wrapper } = setup(mount);

    	expect(wrapper.instance().props.onAddItem).not.toHaveBeenCalled();

    	const formControl = wrapper.find(FormControl).first();
    	formControl.simulate('change', { target: { value: 'a value' } } );
    	wrapper.simulate('submit', 'something');
    	expect(wrapper.instance().props.onAddItem).toHaveBeenCalledWith('a value');
    });

    it('empties the input box when it is submitted', () => {
    	const { wrapper } = setup(mount);

    	const formControl = wrapper.find(FormControl).first();
    	formControl.simulate('change', { target: { value: 'not empty' } } );
		wrapper.simulate('submit');
		expect(formControl.prop('value')).toBe('');
    });

});
