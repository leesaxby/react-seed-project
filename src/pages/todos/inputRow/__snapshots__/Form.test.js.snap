// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Form renders component 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form-group form-group-lg"
  >
    <input
      aria-label={undefined}
      className="form-control"
      data-test-id="todo-add-item"
      id={undefined}
      onChange={[Function]}
      onKeyPress={undefined}
      placeholder={undefined}
      type="text"
      value=""
    />
  </div>
</form>
`;
